### PR TITLE
Updated get delta x & added exhausting tests

### PR DESCRIPTION
**Pull request by Sniezka1927**

### CODE REVIEW

File: contracts/math/clamm.ral [hunk 1]
```diff
@@ -106,17 +108,21 @@ Contract CLAMM() extends Log(){
...
+            return uints.toU256(deltaXUp)
         } else {
-            return divToToken(
-                nominator,
-                mulUp(sqrtPriceA, sqrtPriceB, SqrtPriceScale)
-            )
+            let denominatorUp = mulUp(sqrtPriceA, sqrtPriceB, SqrtPriceScale)
+            let mut deltaX = uints.bigMul512(nominator, one(SqrtPriceScale))
+            deltaX = uints.bigDiv512(deltaX, denominatorUp, 1)
+            deltaX = uints.bigDiv512(deltaX, one(SqrtPriceScale), 1)
```
Comment (zielvna) [inline contracts/math/clamm.ral:124]: one div will probably work there

File: contracts/math/clamm.ral [hunk 2]
```diff
@@ -106,17 +108,21 @@ Contract CLAMM() extends Log(){
...
-        let nominator = mul(deltaSqrtPrice, liquidity, LiquidityScale)
-        if (roundingUp == true) {
-            return divToTokenUp(
-                    nominator,
-                    mul(sqrtPriceA, sqrtPriceB, SqrtPriceScale)
-                )
+        let nominator = uints.bigMulDiv256(deltaSqrtPrice, liquidity, one(LiquidityScale))
+        if (roundingUp) {
+            let denominator = mul(sqrtPriceA, sqrtPriceB, SqrtPriceScale)
+            let mut deltaXUp = uints.bigMul512(nominator, one(SqrtPriceScale))
```
Comment (zielvna) [inline contracts/math/clamm.ral:114]: i think you should use `bigMulDivUp256` here

File: test/clamm.test.ts [hunk 3]
```diff
@@ -199,18 +199,200 @@ describe('math tests', () => {
...
+      const paramsUp = { args: { ...params, roundingUp: true } }
+      const paramsDown = { args: { ...params, roundingUp: false } }
+      const resultUp = (await clamm.contractInstance.methods.getDeltaX(paramsUp)).returns
+      const resultDown = (await clamm.contractInstance.methods.getDeltaX(paramsDown)).returns
+      // expected: 75884792730156830614567103553061795263351065677581979504561495713443442818879n
+      // received: 75884792730156830614567103553061795263351065677581979478702815696568066130226n
+      expect(resultUp).toEqual(75884792730156830614567103553061795263351065677581979478702815696568066130226n)
+      // expected: 75884792730156830614567103553061795263351065677581979504561495713443442818878n
+      // received: 75884792730156830614567103553061795263351065677581979478702815696568066130226n
+      expect(resultDown).toEqual(75884792730156830614567103553061795263351065677581979478702815696568066130225n)
```
Comment (zielvna) [inline test/clamm.test.ts:298]: test results are off

**Review by zielvna — CHANGES_REQUESTED** [review]

**Review by zielvna — APPROVED** [review]

File: contracts/math/clamm.ral [hunk 4]
```diff
@@ -106,17 +108,21 @@ Contract CLAMM() extends Log(){
...
 
-        let nominator = mul(deltaSqrtPrice, liquidity, LiquidityScale)
-        if (roundingUp == true) {
-            return divToTokenUp(
-                    nominator,
-                    mul(sqrtPriceA, sqrtPriceB, SqrtPriceScale)
-                )
+        let nominator = uints.bigMulDiv256(deltaSqrtPrice, liquidity, one(LiquidityScale))
+        if (roundingUp) {
+            let denominator = mul(sqrtPriceA, sqrtPriceB, SqrtPriceScale)
```
Comment (wojciech-cichocki) [inline contracts/math/clamm.ral:113]: It seems that `U512` should be used here

File: contracts/math/clamm.ral [hunk 5]
```diff
@@ -106,17 +108,21 @@ Contract CLAMM() extends Log(){
...
         }
 
-        let nominator = mul(deltaSqrtPrice, liquidity, LiquidityScale)
-        if (roundingUp == true) {
-            return divToTokenUp(
-                    nominator,
-                    mul(sqrtPriceA, sqrtPriceB, SqrtPriceScale)
-                )
+        let nominator = uints.bigMulDiv256(deltaSqrtPrice, liquidity, one(LiquidityScale))
+        if (roundingUp) {
```
Comment (wojciech-cichocki) [inline contracts/math/clamm.ral:112]: I think it's worth using abstraction here and add function that divided two numbers and return token amounts

File: test/clamm.test.ts [hunk 6]
```diff
@@ -199,18 +199,200 @@ describe('math tests', () => {
...
+        sqrtPriceB: almostMinSqrtPrice,
+        liquidity: maxLiquidity
+      }
+      const paramsUp = { args: { ...params, roundingUp: true } }
+      const paramsDown = { args: { ...params, roundingUp: false } }
+      const resultUp = (await clamm.contractInstance.methods.getDeltaX(paramsUp)).returns
+      const resultDown = (await clamm.contractInstance.methods.getDeltaX(paramsDown)).returns
+      // expected: 3794315473971847510172532341754979462199874072217062973965311338137066234n
+      // received: 3794315473971847510172532341754979462199874072217062972672351494741127621n
+      expect(resultUp).toEqual(3794315473971847510172532341754979462199874072217062972672351494741127621n)
```
Comment (wojciech-cichocki) [inline test/clamm.test.ts:325]: What are the causes of these differences in results?

**Review by wojciech-cichocki — CHANGES_REQUESTED** [review]

**Review by wojciech-cichocki — APPROVED** [review]

**Pull request merged**